feat(codebuild): add support for test reports

skinny85 opened the pull request:
### Commit Message
feat(codebuild): add support for test reports

This adds support for CodeBuild's test reports feature.
It includes a new Layer 2 class, ReportGroup.

The default project's IAM role has been augmented with permissions
to create and use test result groups whose names begin with the project's name
(as that is what the default report group created when putting a simple name in the
'reports' buildspec section is called).
There is a new boolean property when creating the project,
addCreateReportGroupPermissions,
that can be used to turn off this behavor.

Fixes #7367

### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cebd527847c443319f8b13dc7dae5d7a1b47e9ad
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0b0ee663-44bb-430c-a077-0b0a5d699681%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-codebuild/README.md:
````diff
@@ -245,6 +245,65 @@ any attempt to save more than one will result in an error.
+you have the the `createsReportGroups()` method:
+
+```typescript
+project.createsReportGroups();
````
Maybe turn this into `ReportGroup.grantCreate(project)` or something, to follow the typical naming pattern & make it clear what this does?

skinny85 replied:
Like we talked about, got rid of this completely, in favor of always adding these permissions.

rix0rrr commented on packages/@aws-cdk/aws-codebuild/README.md:
````diff
@@ -245,6 +245,65 @@ any attempt to save more than one will result in an error.
+If you do that, you need to grant the project's role permissions to write reports to that report group:
+
+```typescript
+project.usesReportGroup(reportGroup);
````
Let's just do the other one?

skinny85 replied:
Done.

rix0rrr commented on packages/@aws-cdk/aws-codebuild/README.md:
````diff
@@ -245,6 +245,65 @@ any attempt to save more than one will result in an error.
+```typescript
+project.usesReportGroup(reportGroup);
+// or, this is equivalent:
+reportGroup.grantUse(project);
````
`grantWrite()`?

skinny85 replied:
Changed.

rix0rrr commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -40,6 +42,40 @@ export interface IProject extends IResource, iam.IGrantable, ec2.IConnectable {
+   *
+   * @see https://docs.aws.amazon.com/codebuild/latest/userguide/test-report-group-naming.html
+   */
+  createsReportGroups(): boolean;
```
I guess don't do this. Follow the grant pattern and have it return an `iam.Grant`?

skinny85 replied:
Done.

polothy commented:
This looks awesome! Was just looking to do this today.  🤞 that is merges soon so I can be lazy and not do it myself 😄

skinny85 commented:
@rix0rrr included your comments, would appreciate a quick re-review! (I get that you approved these already, but there are enough changes since the last revision that I think another glance is justified here)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c7bce03307455ac294225b0c8dc066813dbfc8ba
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9d43e0e7-f744-4b08-b401-7949ed9a8bf5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 16c694de366e055ef6a6021af821602d4f83b46a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0ac3a32f-3527-4a4e-beb2-f7949155fddc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 980146fdf3329e3c6da92238a7dfcd20747f3c91
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=97bbae7b-0650-4172-8464-029b2637b53e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-codebuild/README.md:
````diff
@@ -269,6 +269,68 @@ any attempt to save more than one will result in an error.
+```typescript
+const project = new codebuild.Project(this, 'Project', {
+  // ...
+  addCreateReportGroupPermissions: false,
````
This reads clunky to me because of the `addCreate` at the start. How about `reportGroupPermissions`? Or even `automaticReportingPermission` or something?

skinny85 replied:
How about `grantReportGroupPermissions`?

rix0rrr replied:
Okay

skinny85 replied:
👍

rix0rrr commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -519,6 +520,18 @@ export interface CommonProjectProps {
   readonly fileSystemLocations?: IFileSystemLocation[];
+
+  /**
+   * A switch controlling whether to add permissions to this project's role
```
"A switch controller" is superfluous. It's a boolean toggle, of course it's a switch, and of course it controls something :).

```suggestion
   * Add reporting permissions to this project's role
```

rix0rrr replied:
Also, divide your comment blocks into a title and a body please, just like git commits and emails.

skinny85 replied:
Done.

rix0rrr commented on packages/@aws-cdk/aws-codebuild/lib/report-group.ts:
```diff
@@ -0,0 +1,152 @@
+        'codebuild:UpdateReport',
+        'codebuild:BatchPutTestCases',
+      ],
+      resourceArns: [this.reportGroupArn],
```
Hm. Wonder if `CreateReport`/`UpdateReport` need a wildcard UNDER the report group, like log groups/log streams do (`${this.reportGroupArn}:*`). Could be that they do, could be that they don't.

Might be worth a quick confirmation?

skinny85 replied:
I can confirm the `*` is _not_ required: https://docs.aws.amazon.com/codebuild/latest/userguide/test-permissions.html

(I've also created report groups and projects that used them with these permissions, and everything worked correctly - I was able to see the uploaded reports after the builds finished)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 539ff62898fcaff10ba4823a158a10f13a35c76c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4bc29231-792d-4af2-805c-3eff5f0ff4b6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 1263846b6d2b1534c1104e7ba70250bc95e67c0f
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=393d7642-d737-4362-b991-6457f3cf7bd5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c1bce39740bc2f427d5e05609cb6f7bdbd9392c5
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a2325732-c47b-425d-8230-481970e8aa2b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
